Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp

```diff
@@ -391,1286 +2515,359 @@ double BytecodeInterpreter::read_value<double>(ReadonlyBytes data)
-            }
+            parent[root_b] = root_a;
+            rank[root_a]++;
         }
+    };
 
-        return;
-    }
-    // https://webassembly.github.io/spec/core/bikeshed/#exec-memory-init
-    case Instructions::memory_init.value(): {
-        auto& args = instruction.arguments().get<Instruction::MemoryInitArgs>();
-        auto& data_address = configuration.frame().module().datas()[args.data_index.value()];
-        auto& data = *configuration.store().get(data_address);
-        auto memory_address = configuration.frame().module().memories()[args.memory_index.value()];
-        auto memory = configuration.store().get(memory_address);
-        auto count = configuration.value_stack().take_last().to<u32>();
-        auto source_offset = configuration.value_stack().take_last().to<u32>();
-        auto destination_offset = configuration.value_stack().take_last().to<u32>();
-
-        Checked<size_t> source_position = source_offset;
-        source_position.saturating_add(count);
-        Checked<size_t> destination_position = destination_offset;
-        destination_position.saturating_add(count);
-        TRAP_IF_NOT(source_position <= data.data().size());
-        TRAP_IF_NOT(destination_position <= memory->data().size());
-
-        if (count == 0)
-            return;
+    HashTable<size_t> stack_forced_roots;
+
+    Vector<Vector<size_t>> live_at_instr;
+    live_at_instr.resize(result.dispatches.size());
+
+    for (size_t i = 0; i < result.dispatches.size(); ++i) {
+        auto& dispatch = result.dispatches[i];
+        auto opcode = dispatch.instruction->opcode();
+        size_t inputs = 0;
+        size_t outputs = 0;
+        Vector<size_t> dependent_ids;
+
+        bool variadic_or_unknown = false;
+
+        switch (opcode.value()) {
+#define M(name, _, ins, outs)                    \
+    case Instructions::name.value():             \
+        if constexpr (ins == -1 || outs == -1) { \
+            variadic_or_unknown = true;          \
+        } else {                                 \
+            inputs = ins;                        \
+            outputs = outs;                      \
+        }                                        \
+        break;
+            ENUMERATE_WASM_OPCODES(M)
+#undef M
+        }
 
-        Instruction::MemoryArgument memarg { 0, 0, args.memory_index };
-        for (size_t i = 0; i < (size_t)count; ++i) {
-            auto value = data.data()[source_offset + i];
-            store_to_memory(configuration, memarg, { &value, sizeof(value) }, destination_offset + i);
+        if (variadic_or_unknown) {
+            for (auto val : value_stack) {
+                auto& value = values.get(val).value();
+                value.uses.append(i);
+                value.last_use = max(value.last_use, i);
+                dependent_ids.append(val);
+                forced_stack_values.append(val);
+                live_at_instr[i].append(val);
+            }
```

Comment:
Maybe pre-allocate here, not sure

Reply:
eh, rare enough not to matter much I think.